File: solutions/elixir/109-contains-duplicate.ex
# Problem 109: Contains Duplicate (LeetCode 217)
# Difficulty: Easy
# Language: Elixir
# 
defmodule ContainsDup do
  def check(nums) do
    Enum.reduce_while(nums, MapSet.new(), fn n, seen ->
      if MapSet.member?(seen, n), do: {:halt, true},
      else: {:cont, MapSet.put(seen, n)}
    end) == true
  end
end

File: solutions/rust/109-contains-duplicate.rs
/**
 * Problem 109: Contains Duplicate (LeetCode 217)
 * Difficulty: Easy
 * Language: Rust
 */
use std::collections::HashSet;

fn contains_duplicate(nums: &[i32]) -> bool {
    nums.iter().try_fold(HashSet::new(), |mut s, &n| {
        if s.contains(&n) { Err(()) } else { s.insert(n); Ok(s) }
    }).is_err()
}
